### PR TITLE
Rename options

### LINKED ISSUE #32: Plan the expansion of the options
We are running out of short options and hotkeys, Let's plan how to tackle that.

A few possibilities:

- Introduce modifies: Ctrl, Alt, Shift
- Use vim like commands: `:comp=2`

### PR DESCRIPTION
**Pull request by mwestphal**
Fix #32

### CODE REVIEW

File: application/F3DOptionsParser.cxx [hunk 1]
```diff
@@ -213,72 +215,72 @@ void ConfigurationOptions::GetOptionsFromArgs(
...
-    this->DeclareOption(grp1, "trackball", "k", "Enable trackball interaction", options.getAsBoolRef("trackball"), HasDefault::YES, MayHaveConfig::YES);
-    this->DeclareOption(grp1, "animation-index", "", "Select animation to show", options.getAsIntRef("animation-index"), HasDefault::YES, MayHaveConfig::YES, "<index>");
-    this->DeclareOption(grp1, "font-file", "", "Path to a FreeType compatible font file", options.getAsStringRef("font-file"), HasDefault::NO, MayHaveConfig::NO, "<file_path>");
+    this->DeclareOption(grp1, "progress", "", "Show progress bar", options.getAsBoolRef("loader.show-progress"), HasDefault::YES, MayHaveConfig::YES);
+    this->DeclareOption(grp1, "geometry-only", "", "Do not read materials, cameras and lights from file", options.getAsBoolRef("loader.geometry-only"), HasDefault::YES, MayHaveConfig::YES);
+    this->DeclareOption(grp1, "up", "", "Up direction", options.getAsStringRef("loader.up-direction"), HasDefault::YES, MayHaveConfig::YES, "[-X|+X|-Y|+Y|-Z|+Z]"); // TODO add a config test, critical
```
Comment (Meakk) [inline application/F3DOptionsParser.cxx:220]: will be removed with the creation of the camera

File: application/F3DOptionsParser.cxx [hunk 2]
```diff
@@ -213,72 +215,72 @@ void ConfigurationOptions::GetOptionsFromArgs(
...
+    this->DeclareOption(grp1, "axis", "x", "Show axes", options.getAsBoolRef("interactor.axis"), HasDefault::YES, MayHaveConfig::YES);
+    this->DeclareOption(grp1, "grid", "g", "Show grid", options.getAsBoolRef("window.actor.grid"), HasDefault::YES, MayHaveConfig::YES);
```
Comment (Meakk) [inline application/F3DOptionsParser.cxx:222]: Maybe just `window.grid`?

File: application/F3DOptionsParser.cxx [hunk 3]
```diff
@@ -213,72 +215,72 @@ void ConfigurationOptions::GetOptionsFromArgs(
...
+    this->DeclareOption(grp1, "edges", "e", "Show cell edges", options.getAsBoolRef("window.actor.edges"), HasDefault::YES, MayHaveConfig::YES);
```
Comment (Meakk) [inline application/F3DOptionsParser.cxx:223]: not sure if it's `window` or `loader`

File: application/F3DOptionsParser.cxx [hunk 4]
```diff
@@ -213,72 +215,72 @@ void ConfigurationOptions::GetOptionsFromArgs(
...
-    this->DeclareOption(grp2, "texture-emissive", "", "Path to a texture file that sets the emitted light of the object", options.getAsStringRef("texture-emissive"), HasDefault::NO, MayHaveConfig::YES, "<file path>");
-    this->DeclareOption(grp2, "emissive-factor", "", "Emissive factor. This value is multiplied with the emissive color when an emissive texture is present", options.getAsDoubleVectorRef("emissive-factor"), HasDefault::YES, MayHaveConfig::YES, "<R,G,B>");
-    this->DeclareOption(grp2, "texture-normal", "", "Path to a texture file that sets the normal map of the object", options.getAsStringRef("texture-normal"), HasDefault::NO, MayHaveConfig::YES, "<file path>");
-    this->DeclareOption(grp2, "normal-scale", "", "Normal scale affects the strength of the normal deviation from the normal texture", options.getAsDoubleRef("normal-scale"), HasDefault::YES, MayHaveConfig::YES, "<normalScale>");
+    this->DeclareOption(grp2, "point-sprites", "o", "Show sphere sprites instead of geometry", options.getAsBoolRef("window.point-sprites.enable"), HasDefault::YES, MayHaveConfig::YES);
```
Comment (Meakk) [inline application/F3DOptionsParser.cxx:230]: here too, maybe `loader` is better

File: application/F3DOptionsParser.cxx [hunk 5]
```diff
@@ -213,72 +215,72 @@ void ConfigurationOptions::GetOptionsFromArgs(
...
+    this->DeclareOption(grp2, "point-size", "", "Point size when showing vertices or point sprites", options.getAsDoubleRef("loader.default.point-size"), HasDefault::YES, MayHaveConfig::YES, "<size>");
```
Comment (Meakk) [inline application/F3DOptionsParser.cxx:231]: `loader.default` is not very self-explanatory. What do you think of `loader.material.point-size`? And change all other related options.

File: application/F3DOptionsParser.cxx [hunk 6]
```diff
@@ -213,72 +215,72 @@ void ConfigurationOptions::GetOptionsFromArgs(
...
+    this->DeclareOption(grp2, "line-width", "", "Line width when showing edges", options.getAsDoubleRef("loader.default.line-width"), HasDefault::YES, MayHaveConfig::YES, "<width>");
+    this->DeclareOption(grp2, "color", "", "Solid color", options.getAsDoubleVectorRef("loader.default.color.rgb"), HasDefault::YES, MayHaveConfig::YES, "<R,G,B>");
```
Comment (Meakk) [inline application/F3DOptionsParser.cxx:233]: Maybe just `loader.material.color` here

File: application/F3DOptionsParser.cxx [hunk 7]
```diff
@@ -213,72 +215,72 @@ void ConfigurationOptions::GetOptionsFromArgs(
...
+    this->DeclareOption(grp2, "opacity", "", "Opacity", options.getAsDoubleRef("loader.default.color.opacity"), HasDefault::YES, MayHaveConfig::YES, "<opacity>");
```
Comment (Meakk) [inline application/F3DOptionsParser.cxx:234]: `loader.material.opacity` ?

File: application/F3DOptionsParser.cxx [hunk 8]
```diff
@@ -213,72 +215,72 @@ void ConfigurationOptions::GetOptionsFromArgs(
...
+    this->DeclareOption(grp2, "roughness", "", "Roughness coefficient (0.0-1.0)", options.getAsDoubleRef("loader.default.material.roughness"), HasDefault::YES, MayHaveConfig::YES, "<roughness>");
```
Comment (Meakk) [inline application/F3DOptionsParser.cxx:235]: `loader.material.roughness` ?

File: application/F3DOptionsParser.cxx [hunk 9]
```diff
@@ -213,72 +215,72 @@ void ConfigurationOptions::GetOptionsFromArgs(
...
+    this->DeclareOption(grp2, "metallic", "", "Metallic coefficient (0.0-1.0)", options.getAsDoubleRef("loader.default.material.metallic"), HasDefault::YES, MayHaveConfig::YES, "<metallic>");
```
Comment (Meakk) [inline application/F3DOptionsParser.cxx:236]: `loader.material.metallic` ?

File: application/F3DOptionsParser.cxx [hunk 10]
```diff
@@ -213,72 +215,72 @@ void ConfigurationOptions::GetOptionsFromArgs(
...
+    this->DeclareOption(grp2, "hdri", "", "Path to an image file that will be used as a light source", options.getAsStringRef("window.background.hdri"), HasDefault::NO, MayHaveConfig::YES, "<file path>");
```
Comment (Meakk) [inline application/F3DOptionsParser.cxx:237]: `window.hdri` is enough

File: application/F3DOptionsParser.cxx [hunk 11]
```diff
@@ -213,72 +215,72 @@ void ConfigurationOptions::GetOptionsFromArgs(
...
+    this->DeclareOption(grp2, "texture-base-color", "", "Path to a texture file that sets the color of the object", options.getAsStringRef("loader.default.color.texture"), HasDefault::NO, MayHaveConfig::YES, "<file path>");
```
Comment (Meakk) [inline application/F3DOptionsParser.cxx:238]: `loader.material.tex.color` ?

File: application/F3DOptionsParser.cxx [hunk 12]
```diff
@@ -213,72 +215,72 @@ void ConfigurationOptions::GetOptionsFromArgs(
...
+    this->DeclareOption(grp2, "texture-material", "", "Path to a texture file that sets the Occlusion, Roughness and Metallic values of the object", options.getAsStringRef("loader.default.material.texture"), HasDefault::NO, MayHaveConfig::YES, "<file path>");
```
Comment (Meakk) [inline application/F3DOptionsParser.cxx:239]: `loader.material.tex.orm` ?

File: application/F3DOptionsParser.cxx [hunk 13]
```diff
@@ -213,72 +215,72 @@ void ConfigurationOptions::GetOptionsFromArgs(
...
+    this->DeclareOption(grp2, "texture-emissive", "", "Path to a texture file that sets the emitted light of the object", options.getAsStringRef("loader.default.emissive.texture"), HasDefault::NO, MayHaveConfig::YES, "<file path>");
```
Comment (Meakk) [inline application/F3DOptionsParser.cxx:240]: `loader.material.tex.emissive` ?

File: application/F3DOptionsParser.cxx [hunk 14]
```diff
@@ -213,72 +215,72 @@ void ConfigurationOptions::GetOptionsFromArgs(
...
+    this->DeclareOption(grp2, "emissive-factor", "", "Emissive factor. This value is multiplied with the emissive color when an emissive texture is present", options.getAsDoubleVectorRef("loader.default.emissive.factor"), HasDefault::YES, MayHaveConfig::YES, "<R,G,B>");
```
Comment (Meakk) [inline application/F3DOptionsParser.cxx:241]: `loader.material.emissive` ?

Comment (Meakk) [reply application/F3DOptionsParser.cxx:241]: In any case, we should unify. This emissive factor should act just like other uniform coefficient like color/opacity/roughness/metallic

File: application/F3DOptionsParser.cxx [hunk 15]
```diff
@@ -213,72 +215,72 @@ void ConfigurationOptions::GetOptionsFromArgs(
...
+    this->DeclareOption(grp2, "texture-normal", "", "Path to a texture file that sets the normal map of the object", options.getAsStringRef("loader.default.normal.texture"), HasDefault::NO, MayHaveConfig::YES, "<file path>");
```
Comment (Meakk) [inline application/F3DOptionsParser.cxx:242]: `loader.material.tex.normal` ?

File: application/F3DOptionsParser.cxx [hunk 16]
```diff
@@ -213,72 +215,72 @@ void ConfigurationOptions::GetOptionsFromArgs(
...
+    this->DeclareOption(grp2, "normal-scale", "", "Normal scale affects the strength of the normal deviation from the normal texture", options.getAsDoubleRef("loader.default.normal.scale"), HasDefault::YES, MayHaveConfig::YES, "<normalScale>");
```
Comment (Meakk) [inline application/F3DOptionsParser.cxx:243]: Here as well, the scale is a multiplier, just like other parameters.

File: application/F3DOptionsParser.cxx [hunk 17]
```diff
@@ -213,72 +215,72 @@ void ConfigurationOptions::GetOptionsFromArgs(
...
     this->DeclareOption(grp3, "resolution", "", "Window resolution", appOptions.Resolution, HasDefault::YES, MayHaveConfig::NO, "<width,height>");
-    this->DeclareOption(grp3, "fps", "z", "Display frame per second", options.getAsBoolRef("fps"), HasDefault::YES, MayHaveConfig::YES);
-    this->DeclareOption(grp3, "filename", "n", "Display filename", options.getAsBoolRef("filename"), HasDefault::YES, MayHaveConfig::YES);
-    this->DeclareOption(grp3, "metadata", "m", "Display file metadata", options.getAsBoolRef("metadata"), HasDefault::YES, MayHaveConfig::YES);
-    this->DeclareOption(grp3, "fullscreen", "f", "Full screen", options.getAsBoolRef("fullscreen"), HasDefault::YES, MayHaveConfig::YES);
-    this->DeclareOption(grp3, "blur-background", "u", "Blur background", options.getAsBoolRef("blur-background"), HasDefault::YES, MayHaveConfig::YES);
+    this->DeclareOption(grp3, "fps", "z", "Display frame per second", options.getAsBoolRef("window.actor.fps"), HasDefault::YES, MayHaveConfig::YES);
```
Comment (Meakk) [inline application/F3DOptionsParser.cxx:248]: fps is not related to the actor

**Review by Meakk — CHANGES_REQUESTED** [review]
Ok I'll stop there, I'm very confused.  
Can you provide the list of options in JSON format, like:
```json
{
  "window": {
    "background": {
      "color": [1,1,1],
      "hdri": null
    },
    "actor": {
        ....
    }
  }
}
```

I think it would be easier to see the organization of options this way.

File: application/F3DOptionsParser.cxx (hunk 1 above)
Comment (mwestphal) [reply application/F3DOptionsParser.cxx:220]: nope, camera up and up are two different options

**Comment by mwestphal** [comment]
@Meakk looks at options.cxx instead, it is very clear.

File: library/src/options.cxx [hunk 18]
```diff
@@ -106,63 +106,79 @@ class options::internals
...
-  this->Internals->init("camera-azimuth-angle", 0.0);
-  this->Internals->init("camera-elevation-angle", 0.0);
-  this->Internals->init("camera-focal-point", std::vector<double>());
-  this->Internals->init("camera-position", std::vector<double>());
-  this->Internals->init("camera-view-angle", 0.0);
-  this->Internals->init("camera-view-up", std::vector<double>());
-  this->Internals->init("cells", false);
-  this->Internals->init("colormap",
+  // Loader
+  this->Internals->init("loader.animation.index", 0);
```
Comment (mwestphal) [inline library/src/options.cxx:110]: @Meakk

File: application/F3DOptionsParser.cxx (hunk 14 above)
Comment (mwestphal) [reply application/F3DOptionsParser.cxx:241]: this seems to be already the case

File: application/F3DOptionsParser.cxx (hunk 17 above)
Comment (mwestphal) [reply application/F3DOptionsParser.cxx:248]: it is related, it is an actor we show in the scene.

File: application/F3DOptionsParser.cxx (hunk 4 above)
Comment (mwestphal) [reply application/F3DOptionsParser.cxx:230]: no, as this option is not related to the loader, but to the rendering.

File: application/F3DOptionsParser.cxx (hunk 3 above)
Comment (mwestphal) [reply application/F3DOptionsParser.cxx:223]: it is a rendering related option

File: library/src/options.cxx [hunk 19]
```diff
@@ -106,57 +106,79 @@ class options::internals
...
+
+  this->Internals->init("loader.default.emissive.factor", std::vector<double>{ 1., 1., 1. });
+  this->Internals->init("loader.default.emissive.texture", std::string());
+
+  this->Internals->init("loader.default.normal.texture", std::string());
+  this->Internals->init("loader.default.normal.scale", 1.0);
+
+  this->Internals->init("loader.default.material.metallic", 0.0);
+  this->Internals->init("loader.default.material.roughness", 0.3);
+  this->Internals->init("loader.default.material.texture", std::string());
```
Comment (Meakk) [inline library/src/options.cxx:132]: I'd prefer `loader.material.xxxx` instead of `loader.default.xxx`.  
And for the last ones:
- `loader.default.material.metallic` -> `loader.material.pbr.metallic`
- `loader.default.material.roughness` -> `loader.material.pbr.roughness`
- `loader.default.material.texture` -> `loader.material.pbr.texture`

File: library/src/options.cxx [hunk 20]
```diff
@@ -106,57 +106,79 @@ class options::internals
...
+  this->Internals->init("window.background.blur", false);
+
+  this->Internals->init("window.camera.azimuth-angle", 0.0);
+  this->Internals->init("window.camera.elevation-angle", 0.0);
+  this->Internals->init("window.camera.focal-point", std::vector<double>());
+  this->Internals->init("window.camera.position", std::vector<double>());
+  this->Internals->init("window.camera.view-angle", 0.0);
+  this->Internals->init("window.camera.view-up", std::vector<double>());
+
+  this->Internals->init("window.coloring.cells", false);
```
Comment (Meakk) [inline library/src/options.cxx:147]: I think those are related to `loader`, it's just sciviz material parameters.

File: library/src/options.cxx [hunk 21]
```diff
@@ -106,57 +106,79 @@ class options::internals
...
+  this->Internals->init("window.raytracing.enable", false);
+  this->Internals->init("window.raytracing.denoise", false);
+  this->Internals->init("window.raytracing.samples", 5);
+  this->Internals->init("window.actor.bar", false);
+  this->Internals->init("window.actor.edges", false);
+  this->Internals->init("window.actor.filename", false);
+  this->Internals->init("window.actor.fps", false);
+  this->Internals->init("window.actor.grid", false);
+  this->Internals->init("window.actor.cheatsheet", false);
+  this->Internals->init("window.actor.metadata", false);
```
Comment (Meakk) [inline library/src/options.cxx:164]: These are not related to the actor in my opinion, what about `window.display.xxxx`

File: library/src/options.cxx [hunk 22]
```diff
@@ -106,57 +106,79 @@ class options::internals
...
+
+  this->Internals->init("window.render.depth-peeling", false);
+  this->Internals->init("window.render.fxaa", false);
+  this->Internals->init("window.render.ssao", false);
+  this->Internals->init("window.render.tone-mapping", false);
```
Comment (Meakk) [inline library/src/options.cxx:169]: `window.effect`?

File: library/src/options.cxx [hunk 23]
```diff
@@ -106,57 +106,79 @@ class options::internals
...
+
+  this->Internals->init("window.fullscreen", false);
```
Comment (Meakk) [inline library/src/options.cxx:171]: to remove

**Review by Meakk — CHANGES_REQUESTED** [review]

File: library/src/options.cxx [hunk 24]
```diff
@@ -106,57 +106,79 @@ class options::internals
...
-  this->Internals->init("axis", false);
-  this->Internals->init("background-color", std::vector<double>{ 0.2, 0.2, 0.2 });
-  this->Internals->init("bar", false);
-  this->Internals->init("blur-background", false);
-  this->Internals->init("cells", false);
-  this->Internals->init("colormap",
+  // Loader
+  this->Internals->init("loader.animation.index", 0);
+  this->Internals->init("loader.camera.index", -1);
+  this->Internals->init("loader.up-direction", std::string("+Y"));
```
[the remark this answers is not in the document]
Comment (mwestphal) [reply library/src/options.cxx:112]: sadly it needs to be defined at load time, once it is part of the loader options.

File: library/src/options.cxx (hunk 19 above)
Comment (mwestphal) [reply library/src/options.cxx:132]: how do user know at first glance that an option is specific to the default scene then ?

File: library/src/options.cxx (hunk 20 above)
Comment (mwestphal) [reply library/src/options.cxx:147]: it can be changed dynamically in the window, it is related to rendering.

File: library/src/options.cxx (hunk 21 above)
Comment (mwestphal) [reply library/src/options.cxx:164]: that would be fine, much better actually

File: library/src/options.cxx (hunk 23 above)
Comment (mwestphal) [reply library/src/options.cxx:171]: the option is still present, just no more in the hotkeys

File: library/src/options.cxx (hunk 22 above)
Comment (Meakk) [reply library/src/options.cxx:169]: yes

File: library/src/options.cxx (hunk 19 above)
Comment (mwestphal) [reply library/src/options.cxx:132]: and why adding pbr ?
do you plan having non-pbr materials ?

Comment (Meakk) [reply library/src/options.cxx:132]: From the documentation, the user knows that using the default scene means the material is not loaded.

File: library/src/options.cxx (hunk 24 above)
[the remark this answers is not in the document]
Comment (Meakk) [reply library/src/options.cxx:112]: It should be be the case. We can easily have a hotkey to change the orientation if we want to.

[the remark this answers is not in the document]
Comment (mwestphal) [reply library/src/options.cxx:112]: will be reworked

File: library/src/options.cxx (hunk 19 above)
Comment (mwestphal) [reply library/src/options.cxx:132]: this may not be enough imo

File: library/src/options.cxx (hunk 20 above)
Comment (mwestphal) [reply library/src/options.cxx:147]: will be reworked

File: application/F3DOptionsParser.cxx (hunk 10 above)
Comment (mwestphal) [reply application/F3DOptionsParser.cxx:237]: I kept the background, let me know

**Review by Meakk — APPROVED** [review]
